src/deploy/functions/release/index.ts

```diff
@@ -144,3 +141,119 @@ export function printTriggerUrls(results: backend.Backend): void {
     logger.info(clc.bold("Function URL"), `(${getFunctionLabel(httpsFunc)}):`, httpsFunc.uri);
   }
 }
+
+/**
+ * Checks if artifact cleanup policies are set for the regions where functions are deployed
+ * and automatically sets up policies where needed.
+ *
+ * The policy is only set up when:
+ * 1. No cleanup policy exists yet
+ * 2. No other cleanup policies exist (beyond our own if we previously set one)
+ * 3. User has not explicitly opted out
+ */
+async function checkArtifactCleanupPolicies(
+  projectId: string,
+  endpoints: backend.Endpoint[],
+): Promise<void> {
+  if (endpoints.length === 0) {
+    return;
+  }
+
+  const uniqueRegions = new Set<string>();
+  for (const endpoint of endpoints) {
+    uniqueRegions.add(endpoint.region);
+  }
+
+  const regionResults = await Promise.all(
+    Array.from(uniqueRegions).map(async (region) => {
+      try {
+        const repoPath = artifacts.makeRepoPath(projectId, region);
+        const repository = await artifactregistry.getRepository(repoPath);
+        const existingPolicy = artifacts.findExistingPolicy(repository);
+        const hasPolicy = !!existingPolicy;
+        const hasOptOut = artifacts.hasCleanupOptOut(repository);
+
+        // Check if there are any other cleanup policies beyond our own
+        const hasOtherPolicies =
+          repository.cleanupPolicies &&
+          Object.keys(repository.cleanupPolicies).some(
+            (key) => key !== artifacts.CLEANUP_POLICY_ID,
+          );
+
+        return {
+          region,
+          repository,
+          hasPolicy,
+          hasOptOut,
+          hasOtherPolicies,
+        };
+      } catch (err) {
+        logger.debug(`Failed to check artifact cleanup policy for region ${region}:`, err);
+        return {
+          region,
+          hasPolicy: false,
+          hasOptOut: false,
+          hasOtherPolicies: false,
+          error: err,
+        };
+      }
+    }),
+  );
+
+  const regionsWithErrors = regionResults
```

Comment:
It might be more idiomatic to use `Promise.allSettled()` and then filter out the resolved vs rejected promises. Here, `Promise.all()` is going to fail on the first error and you'll get effectively undefined behavior.